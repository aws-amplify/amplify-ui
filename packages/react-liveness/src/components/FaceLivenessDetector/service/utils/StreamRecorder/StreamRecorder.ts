import { isUndefined } from '@aws-amplify/ui';
import { TIME_SLICE } from '../constants';
import { VideoStream, StreamResult, StreamResultType } from '../types';

export class StreamRecorder {
  #chunks: Blob[];
  #recorder: MediaRecorder;
  #recordingStarted: boolean = false;
  #recorderEndTimestamp: number | undefined;
  #recorderStartTimestamp: number | undefined;
  #recordingStartTimestamp: number | undefined;
  #recorderStopped!: Promise<void>;
  #videoStream: VideoStream;

  constructor(stream: MediaStream) {
    if (typeof MediaRecorder === 'undefined') {
      throw Error('MediaRecorder is not supported by this browser');
    }

    this.#chunks = [];
    this.#recorder = new MediaRecorder(stream, { bitsPerSecond: 1000000 });
    this.#videoStream = this.#createReadableStream();

    this.#setupCallbacks();
  }

  getVideoStream(): VideoStream {
    return this.#videoStream;
  }

  dispatchStreamEvent<T extends StreamResultType>(
    event: T extends 'streamStop'
      ? Pick<StreamResult<T>, 'type'>
      : StreamResult<T>
  ): void {
    const { type } = event;
    const data = type === 'streamStop' ? undefined : event.data;
    this.#recorder.dispatchEvent(new MessageEvent(type, { data }));
  }

  getRecordingStartTimestamp(): number {
    if (
      isUndefined(this.#recorderStartTimestamp) ||
      isUndefined(this.#recordingStartTimestamp)
    ) {
      throw new Error('Recording has not started');
    }
    /**
     * This calculation is provided by Science team after doing analysis
     * of unreliable .onstart() (this.#recorderStartTimestamp) timestamp that is
     * returned from mediaRecorder.
     */
    return Math.round(
      0.73 * (this.#recorderStartTimestamp - this.#recordingStartTimestamp) +
        this.#recordingStartTimestamp
    );
  }

  getRecordingEndedTimestamp(): number {
    if (isUndefined(this.#recorderEndTimestamp)) {
      throw new Error('Recording has not ended');
    }
    return this.#recorderEndTimestamp;
  }

  startRecording(): void {
    this.#clearRecordedChunks();
    this.#recordingStartTimestamp = Date.now();
    this.#recorder.start(TIME_SLICE);
  }

  isRecording(): boolean {
    return this.#recorder.state === 'recording';
  }

  getChunksLength(): number {
    return this.#chunks.length;
  }

  hasRecordingStarted(): boolean {
    return this.#recordingStarted;
  }

  async stopRecording(): Promise<void> {
    if (this.isRecording()) {
      this.#recorder.stop();
    }
    return this.#recorderStopped;
  }

  #clearRecordedChunks(): void {
    this.#chunks = [];
  }

  // Creates a Readablestream of video chunks. Waits to receive a clientSessionInfo event before pushing
  // a livenessActionDocument to the ReadableStream and finally closing the ReadableStream
  #createReadableStream(): ReadableStream {
    return new ReadableStream<StreamResult>({
      start: (controller) => {
        this.#recorder.ondataavailable = ({ data }) => {
          this.#chunks.push(data);
          controller.enqueue({ type: 'streamVideo', data });
        };

        this.#recorder.addEventListener('sessionInfo', (e: any) => {
          const { data } = e as unknown as StreamResult<'sessionInfo'>;
          controller.enqueue({ type: 'sessionInfo', data });
        });

        this.#recorder.addEventListener('streamStop', () => {
          controller.enqueue({ type: 'streamStop' });
        });

        this.#recorder.addEventListener('closeCode', (e) => {
          const { data } = e as unknown as StreamResult<'closeCode'>;
          controller.enqueue({ type: 'closeCode', data });
        });

        this.#recorder.addEventListener('endStream', () => {
          controller.close();
        });
      },
    });
  }

  #setupCallbacks() {
    this.#recorder.onstart = () => {
      this.#recordingStarted = true;
      this.#recorderStartTimestamp = Date.now();
    };

    this.#recorderStopped = new Promise((resolve) => {
      this.#recorder.onstop = () => {
        this.#recorderEndTimestamp = Date.now();
        resolve();
      };
    });

    this.#recorder.onerror = () => {
      this.stopRecording();
    };
  }
}
