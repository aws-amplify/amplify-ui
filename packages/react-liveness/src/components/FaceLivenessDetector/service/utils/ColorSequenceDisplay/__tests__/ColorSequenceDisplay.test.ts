import 'web-streams-polyfill';
import 'jest-canvas-mock';

import { StreamRecorder } from '../../StreamRecorder';

import {
  ColorSequenceDisplay,
  ColorSequences,
  SequenceColorValue,
  StartSequencesParams,
} from '../ColorSequenceDisplay';

jest.mock('../../StreamRecorder');

// sequence values from service
const colorSequences: ColorSequences = [
  { color: 'rgb(0,0,0)', downscrollDuration: 0, flatDisplayDuration: 75 },
  { color: 'rgb(0,255,0)', downscrollDuration: 475, flatDisplayDuration: 0 },
  { color: 'rgb(255,0,255)', downscrollDuration: 475, flatDisplayDuration: 0 },
  { color: 'rgb(255,0,0)', downscrollDuration: 475, flatDisplayDuration: 0 },
  { color: 'rgb(0,0,255)', downscrollDuration: 475, flatDisplayDuration: 0 },
  { color: 'rgb(255,0,0)', downscrollDuration: 475, flatDisplayDuration: 0 },
  { color: 'rgb(0,255,0)', downscrollDuration: 475, flatDisplayDuration: 0 },
  { color: 'rgb(255,0,0)', downscrollDuration: 475, flatDisplayDuration: 0 },
];

const colorMap: Record<SequenceColorValue, number[]> = {
  'rgb(0,0,0)': [0, 0, 0],
  'rgb(0,255,0)': [0, 255, 0],
  'rgb(255,0,255)': [255, 0, 255],
  'rgb(255,0,0)': [255, 0, 0],
  'rgb(0,0,255)': [0, 0, 255],
};

const challengeId = 'challengeId';
const stream = {} as MediaStream;
const onSequenceColorChange = jest.fn();

describe('ColorSequenceDisplay', () => {
  let mockDispatchStreamEvent: jest.Mock;
  let canvasElement: HTMLCanvasElement;

  let onSequencesComplete: StartSequencesParams['onSequencesComplete'];
  let onSequenceStart: StartSequencesParams['onSequenceStart'];

  let startParams: StartSequencesParams;

  beforeEach(() => {
    let mockTimestamp = 0;
    jest.spyOn(Date, 'now').mockImplementation(() => {
      mockTimestamp = mockTimestamp + 1000;
      return mockTimestamp;
    });

    const streamRecorder = new StreamRecorder(stream);
    mockDispatchStreamEvent = streamRecorder.dispatchStreamEvent as jest.Mock;

    canvasElement = document.createElement('canvas');
    onSequenceStart = jest.fn(() => {
      canvasElement.style.display = 'block';
    });
    onSequencesComplete = jest.fn(() => {
      canvasElement.style.display = 'none';
    });
    const onSequenceChange: StartSequencesParams['onSequenceChange'] = ({
      prevSequenceColor,
      sequenceColor,
      sequenceIndex,
      sequenceStartTime,
    }) => {
      const CurrentColor = { RGB: colorMap[sequenceColor] };
      const PreviousColor = {
        RGB: colorMap[prevSequenceColor],
      };
      const data = {
        Challenge: {
          FaceMovementAndLightChallenge: {
            ChallengeId: challengeId,
            ColorDisplayed: {
              CurrentColor,
              PreviousColor,
              SequenceNumber: sequenceIndex,
              CurrentColorStartTimestamp: sequenceStartTime,
            },
          },
        },
      };

      mockDispatchStreamEvent({ type: 'sessionInfo', data });
    };

    onSequenceColorChange.mockClear();

    startParams = {
      onSequenceChange,
      onSequenceColorChange,
      onSequenceStart,
      onSequencesComplete,
    };
  });

  it('progresses through expected sequences as expected', async () => {
    const display = new ColorSequenceDisplay(colorSequences);

    expect(canvasElement.style.display).toBe('');

    // first sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(2);
    // first and second dispatches happen call on first call of `start`
    expect(mockDispatchStreamEvent.mock.calls[0][0]).toMatchSnapshot();
    // second call
    expect(mockDispatchStreamEvent.mock.calls[1][0]).toMatchSnapshot();
    expect(onSequenceStart).toBeCalledTimes(1);
    expect(canvasElement.style.display).toBe('block');
    expect(onSequenceColorChange).toBeCalledTimes(1);
    expect(onSequenceColorChange).toBeCalledWith({
      heightFraction: 0,
      sequenceColor: 'rgb(0,255,0)',
      prevSequenceColor: 'rgb(0,0,0)',
    });

    // second sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(3);
    expect(mockDispatchStreamEvent.mock.calls[2][0]).toMatchSnapshot();
    expect(canvasElement.style.display).toBe('block');

    // third sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(4);
    expect(mockDispatchStreamEvent.mock.calls[3][0]).toMatchSnapshot();
    expect(canvasElement.style.display).toBe('block');

    // fourth sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(5);
    expect(mockDispatchStreamEvent.mock.calls[4][0]).toMatchSnapshot();
    expect(canvasElement.style.display).toBe('block');

    // fifth sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(6);
    expect(mockDispatchStreamEvent.mock.calls[5][0]).toMatchSnapshot();
    expect(canvasElement.style.display).toBe('block');

    // sixth sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(7);
    expect(mockDispatchStreamEvent.mock.calls[6][0]).toMatchSnapshot();
    expect(canvasElement.style.display).toBe('block');

    // seventh sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(8);
    expect(mockDispatchStreamEvent.mock.calls[7][0]).toMatchSnapshot();
    expect(onSequenceStart).toBeCalledTimes(7);
    expect(canvasElement.style.display).toBe('block');

    // eighth sequence
    expect(await display.startSequences(startParams)).toBe(true);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(8);
    expect(onSequenceStart).toBeCalledTimes(8);
    expect(onSequencesComplete).toHaveBeenCalledTimes(1);
    expect(canvasElement.style.display).toBe('none');
  });

  it('handles a sequence that updates `ColorStageType` from `SCROLLING` to `FLAT` as expected', async () => {
    const initialSequence = colorSequences[0];
    const scrollingSequence = colorSequences[1];
    const flatSequence = colorSequences[0];

    const display = new ColorSequenceDisplay([
      initialSequence,
      scrollingSequence,
      flatSequence,
    ]);

    // first sequence
    expect(await display.startSequences(startParams)).toBe(false);
    // first and second dispatches happen call on first call of `start`
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(2);

    // second sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(3);

    // third sequence
    expect(await display.startSequences(startParams)).toBe(false);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(4);

    // third sequence
    expect(await display.startSequences(startParams)).toBe(true);
    expect(mockDispatchStreamEvent).toHaveBeenCalledTimes(4);
  });
});
