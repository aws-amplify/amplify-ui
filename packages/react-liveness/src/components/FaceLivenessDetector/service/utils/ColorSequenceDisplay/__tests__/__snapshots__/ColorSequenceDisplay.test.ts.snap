// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSequenceDisplay progresses through expected sequences as expected 1`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              0,
              0,
              0,
            ],
          },
          "CurrentColorStartTimestamp": 1000,
          "PreviousColor": {
            "RGB": [
              0,
              0,
              0,
            ],
          },
          "SequenceNumber": 0,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 2`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              0,
              255,
              0,
            ],
          },
          "CurrentColorStartTimestamp": 1000,
          "PreviousColor": {
            "RGB": [
              0,
              0,
              0,
            ],
          },
          "SequenceNumber": 1,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 3`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              255,
              0,
              255,
            ],
          },
          "CurrentColorStartTimestamp": 4000,
          "PreviousColor": {
            "RGB": [
              0,
              255,
              0,
            ],
          },
          "SequenceNumber": 2,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 4`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              255,
              0,
              0,
            ],
          },
          "CurrentColorStartTimestamp": 6000,
          "PreviousColor": {
            "RGB": [
              255,
              0,
              255,
            ],
          },
          "SequenceNumber": 3,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 5`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              0,
              0,
              255,
            ],
          },
          "CurrentColorStartTimestamp": 8000,
          "PreviousColor": {
            "RGB": [
              255,
              0,
              0,
            ],
          },
          "SequenceNumber": 4,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 6`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              255,
              0,
              0,
            ],
          },
          "CurrentColorStartTimestamp": 10000,
          "PreviousColor": {
            "RGB": [
              0,
              0,
              255,
            ],
          },
          "SequenceNumber": 5,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 7`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              0,
              255,
              0,
            ],
          },
          "CurrentColorStartTimestamp": 12000,
          "PreviousColor": {
            "RGB": [
              255,
              0,
              0,
            ],
          },
          "SequenceNumber": 6,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;

exports[`ColorSequenceDisplay progresses through expected sequences as expected 8`] = `
{
  "data": {
    "Challenge": {
      "FaceMovementAndLightChallenge": {
        "ChallengeId": "challengeId",
        "ColorDisplayed": {
          "CurrentColor": {
            "RGB": [
              255,
              0,
              0,
            ],
          },
          "CurrentColorStartTimestamp": 14000,
          "PreviousColor": {
            "RGB": [
              0,
              255,
              0,
            ],
          },
          "SequenceNumber": 7,
        },
      },
    },
  },
  "type": "sessionInfo",
}
`;
